File: trees/problems/543_diameter_of_binary_tree.py
# Given the root of a binary tree, return the length of the diameter of the tree.

# The diameter of a binary tree is the length of the longest path between any two nodes in a tree. This path may or may not pass through the root.

# The length of a path between two nodes is represented by the number of edges between them.

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        
        self.diameter = 0
        def traverse(root):
            if not root:
                return 0
            l = traverse(root.left)
            r = traverse(root.right)
            self.diameter = max(self.diameter, l + r)
            return max(l, r) + 1
        traverse(root)
        return self.diameter

File: typescript/problems/543_diameter_of_binary_tree.ts
// Given the root of a binary tree, return the length of the diameter of the tree.

// The diameter of a binary tree is the length of the longest path between any two nodes in a tree. This path may or may not pass through the root.

// The length of a path between two nodes is represented by the number of edges between them.

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function diameterOfBinaryTree(root: TreeNode | null): number {
    let ans: number = 0;
    function dfs(n: TreeNode): number {
        if (n === null) {
            return 0
        }
        const l = dfs(n.left);
        const r = dfs(n.right);
        ans = Math.max(ans, l+r);
        return Math.max(l, r) + 1
    }
    dfs(root)
    return ans
    
};